### PR TITLE
Dual Execution

### PR DESCRIPTION
**Pull request by sinui0**
Here is work on the implementation of Dual Execution with the leaky equality check

### CODE REVIEW

File: mpc-core/src/garble/execution/dual.rs [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+    },
+    utils::{choose, sha256},
+    Block,
+};
+use mpc_circuits::Circuit;
+
+use aes::{Aes128, NewBlockCipher};
+use std::{marker::PhantomData, sync::Arc};
+
+#[derive(Clone, PartialEq)]
```
Comment (th4s) [inline mpc-core/src/garble/execution/dual.rs:15]: Add `Debug` and `Copy`?

**Review by th4s — COMMENTED** [review]
Great work :), some thoughts

- we sometimes have to do with `Vec<Vec<T>>`, sometimes I see `Vec<[T; N]>` or `[[T;N];M]`. Do we have some guiding principles on when to use what when dealing with 2d collections.
- In `garble/evaluator.rs` there are a lot of functions which I would have expected to be methods of some struct. Does if offer any advantages to have these functions not associated to some struct and implement the std `And, Or, Xor, ...` traits

Considering the GCDE typestate pattern, I really like this approach and think it prevents a lot of `if` and `Option<T>`. What do you think about reducing the signature complexity a little bit by not having a `Role` trait, and just have separate `Generator` and `Evaluator` structs, which use the `State` traits.

File: mpc-core/src/garble/circuit.rs [hunk 2]
```diff
@@ -16,6 +18,12 @@ pub struct WireLabel<T> {
     pub value: T,
 }
 
+impl<T> WireLabel<T> {
+    pub fn value(&self) -> &T {
```
[the remark this answers is not in the document]
Comment (sinui0) [reply mpc-core/src/garble/circuit.rs]: Good call, made this change.

File: mpc-core/src/garble/circuit.rs [hunk 3]
```diff
@@ -25,41 +33,147 @@ impl WireLabel<Block> {
...
 
+#[derive(Debug, Clone)]
+pub struct InputValue {
+    input: Input,
+    value: Vec<bool>,
+}
+
+impl InputValue {
+    pub fn new(input: Input, value: &[bool]) -> Self {
+        assert!(input.group().len() == value.len());
```
[the remark this answers is not in the document]
Comment (sinui0) [reply mpc-core/src/garble/circuit.rs:49]: yeah, I just threw this in quick, needs to be handled properly.

**Comment by sinui0** [comment]
> Do we have some guiding principles on when to use what when dealing with 2d collections.

I think you're referring to how we use `Vec<[Block; 2]>` in many places? In these cases it will always be `[T; 2]`, as it denotes label pairs and the accessor syntax is nicer than `(T, T)`. If we use a tuple container then we have `if b { a.1 } else { a.0 }` all over the place instead of `a[b as usize]`. Although I do not like having to cast to usize either. I'd be willing to use tuples instead if argued for

I haven't thought too much about in other cases, but I do agree that `Vec<Vec<T>>` should be avoided as you mentioned last week. I did consider using [bitvec](https://docs.rs/bitvec/latest/bitvec/slice/struct.BitSlice.html) for OT, but I was slapping together a prototype with intent to optimize later.

> In garble/evaluator.rs there are a lot of functions which I would have expected to be methods of some struct. Does if offer any advantages to have these functions not associated to some struct and implement the std And, Or, Xor, ... traits

Do you have an example of what you had in mind? I did previously have a `HalfGateEvaluator` struct or something along those lines, but then I realized that there isn't really any state that needs to be persisted so I converted it to functional. The generation and handling of input labels should be decoupled as we will have instances of reusing them across executions. I figured we'd just import those functions into our execution protocols such as how I've done with `DualExecution`. Definitely not set on that approach if you have some ideas to make it more intuitive.

> What do you think about reducing the signature complexity a little bit by not having a Role trait, and just have separate Generator and Evaluator structs, which use the State traits.

Do you mean having separate `Leader` and `Follower` structs which are generic over state? Both roles take turns acting as the `Generator` and `Evaluator`, I figured with this pattern we could reduce duplication where their behavior is the same. For example `Leader` and `Follower` share behavior for the `Generator` and `Complete` (I didnt implement `Complete` yet) states. I would have combined the `Evaluator` state behavior as well, but they compute their output checks differently. Though you are right that it adds a bit of complexity, I could just pull the `Generator` behavior out into a private function and call it. I'll split them up as you suggest and see how it feels.

**Comment by themighty1** [comment]
Nice work, lgtm

**Comment by th4s** [comment]
> I think you're referring to how we use Vec<[Block; 2]> in many places? In these cases...

Agree, would also choose array over tuple in this case

> I haven't thought too much about in other cases, but I do agree that Vec<Vec<T>> should be avoided...

Agree, though I never really benchmarked it ^^

> Do you have an example of what you had in mind?...

Ahh I see now...  So these are mostly stateless functions, which are meant to be reused in other places. Alright then :)

> Do you mean having separate Leader and Follower structs which are generic over state?...

Yes. 

> I could just pull the Generator behavior out into a private function and call it. I'll split them up as you suggest and see how it feels.

Yeah exactly, maybe this gives a nicer interface and then we only have to deal with a single generic.

**Review by themighty1 — CHANGES_REQUESTED** [review]

File: mpc-core/src/garble/circuit.rs [hunk 4]
```diff
@@ -101,18 +217,53 @@ pub fn generate_labels<R: Rng + CryptoRng>(
     };
     let low = Block::random_vec(rng, count);
```
[the remark this answers is not in the document]
Comment (sinui0) [reply mpc-core/src/garble/circuit.rs:223]: It's a carry over from electrical engineering, low is false truth value (indicating a wire at 0v)

**Pull request merged**